File: planner/cardinality/row_count_column.go
// Copyright 2023 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package cardinality

import (
	"github.com/pingcap/errors"
	"github.com/pingcap/tidb/planner/util/debugtrace"
	"github.com/pingcap/tidb/sessionctx"
	"github.com/pingcap/tidb/statistics"
	"github.com/pingcap/tidb/types"
	"github.com/pingcap/tidb/util/ranger"
)

func init() {
	statistics.GetRowCountByColumnRanges = GetRowCountByColumnRanges
	statistics.GetRowCountByIntColumnRanges = GetRowCountByIntColumnRanges
	statistics.GetRowCountByIndexRanges = GetRowCountByIndexRanges
	statistics.OutOfRangeEQSelectivity = outOfRangeEQSelectivity
	statistics.GetEqualCondSelectivity = getEqualCondSelectivity
}

// GetRowCountByColumnRanges estimates the row count by a slice of Range.
func GetRowCountByColumnRanges(sctx sessionctx.Context, coll *statistics.HistColl, colID int64, colRanges []*ranger.Range) (result float64, err error) {
	var name string
	if sctx.GetSessionVars().StmtCtx.EnableOptimizerDebugTrace {
		debugtrace.EnterContextCommon(sctx)
		debugTraceGetRowCountInput(sctx, colID, colRanges)
		defer func() {
			debugtrace.RecordAnyValuesWithNames(sctx, "Name", name, "Result", result)
			debugtrace.LeaveContextCommon(sctx)
		}()
	}
	sc := sctx.GetSessionVars().StmtCtx
	c, ok := coll.Columns[colID]
	recordUsedItemStatsStatus(sctx, c, coll.PhysicalID, colID)
	if c != nil && c.Info != nil {
		name = c.Info.Name.O
	}
	if !ok || c.IsInvalid(sctx, coll.Pseudo) {
		result, err = GetPseudoRowCountByColumnRanges(sc, float64(coll.RealtimeCount), colRanges, 0)
		if err == nil && sc.EnableOptimizerCETrace && ok {
			CETraceRange(sctx, coll.PhysicalID, []string{c.Info.Name.O}, colRanges, "Column Stats-Pseudo", uint64(result))
		}
		return result, err
	}
	if sctx.GetSessionVars().StmtCtx.EnableOptimizerDebugTrace {
		debugtrace.RecordAnyValuesWithNames(sctx,
			"Histogram NotNull Count", c.Histogram.NotNullCount(),
			"TopN total count", c.TopN.TotalCount(),
			"Increase Factor", c.GetIncreaseFactor(coll.RealtimeCount),
		)
	}
	result, err = c.GetColumnRowCount(sctx, colRanges, coll.RealtimeCount, coll.ModifyCount, false)
	if sc.EnableOptimizerCETrace {
		CETraceRange(sctx, coll.PhysicalID, []string{c.Info.Name.O}, colRanges, "Column Stats", uint64(result))
	}
	return result, errors.Trace(err)
}

// GetRowCountByIntColumnRanges estimates the row count by a slice of IntColumnRange.
func GetRowCountByIntColumnRanges(sctx sessionctx.Context, coll *statistics.HistColl, colID int64, intRanges []*ranger.Range) (result float64, err error) {
	var name string
	if sctx.GetSessionVars().StmtCtx.EnableOptimizerDebugTrace {
		debugtrace.EnterContextCommon(sctx)
		debugTraceGetRowCountInput(sctx, colID, intRanges)
		defer func() {
			debugtrace.RecordAnyValuesWithNames(sctx, "Name", name, "Result", result)
			debugtrace.LeaveContextCommon(sctx)
		}()
	}
	sc := sctx.GetSessionVars().StmtCtx
	c, ok := coll.Columns[colID]
	recordUsedItemStatsStatus(sctx, c, coll.PhysicalID, colID)
	if c != nil && c.Info != nil {
		name = c.Info.Name.O
	}
	if !ok || c.IsInvalid(sctx, coll.Pseudo) {
		if len(intRanges) == 0 {
			return 0, nil
		}
		if intRanges[0].LowVal[0].Kind() == types.KindInt64 {
			result = getPseudoRowCountBySignedIntRanges(intRanges, float64(coll.RealtimeCount))
		} else {
			result = getPseudoRowCountByUnsignedIntRanges(intRanges, float64(coll.RealtimeCount))
		}
		if sc.EnableOptimizerCETrace && ok {
			CETraceRange(sctx, coll.PhysicalID, []string{c.Info.Name.O}, intRanges, "Column Stats-Pseudo", uint64(result))
		}
		return result, nil
	}
	if sctx.GetSessionVars().StmtCtx.EnableOptimizerDebugTrace {
		debugtrace.RecordAnyValuesWithNames(sctx,
			"Histogram NotNull Count", c.Histogram.NotNullCount(),
			"TopN total count", c.TopN.TotalCount(),
			"Increase Factor", c.GetIncreaseFactor(coll.RealtimeCount),
		)
	}
	result, err = c.GetColumnRowCount(sctx, intRanges, coll.RealtimeCount, coll.ModifyCount, true)
	if sc.EnableOptimizerCETrace {
		CETraceRange(sctx, coll.PhysicalID, []string{c.Info.Name.O}, intRanges, "Column Stats", uint64(result))
	}
	return result, errors.Trace(err)
}

File: rust/difftests/planner-tests/tests/row_count_column.rs
// Copyright 2026 PingCAP, Inc.
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

//! Dependency-closed tests for `pkg/planner/cardinality/row_count_column.go`.

use tidb_planner::cardinality::{
    pseudo::{PseudoBoundKind, ScalarRange},
    row_count_column::{
        estimate_column_row_count, pseudo_row_count_with_partial_stats, ColumnRangeInput,
        ColumnRangeStats, PartialStatsRange, RowEstimate,
    },
};

#[test]
fn point_ranges_preserve_primary_key_and_scaled_column_rules() {
    let point = ScalarRange::new(7.0, 7.0, PseudoBoundKind::Value, PseudoBoundKind::Value);
    let stats = ColumnRangeStats::point(RowEstimate::default_est(2.0));
    let input = [ColumnRangeInput::new(point, false, false, stats)];

    assert_eq!(
        estimate_column_row_count(&input, 100.0, 100.0, 0.0, 1.5, true, 0.01),
        RowEstimate::default_est(1.0)
    );
    assert_eq!(
        estimate_column_row_count(&input, 100.0, 100.0, 0.0, 1.5, false, 0.01),
        RowEstimate::default_est(3.0)
    );
}

#[test]
fn interval_ranges_adjust_exclusive_boundaries_and_null() {
    let interval = ScalarRange::new(1.0, 10.0, PseudoBoundKind::Value, PseudoBoundKind::Value);
    let stats = ColumnRangeStats::new(
        RowEstimate::default_est(0.0),
        RowEstimate::new(10.0, 8.0, 12.0),
        RowEstimate::new(2.0, 1.0, 3.0),
        RowEstimate::default_est(1.0),
        None,
    );
    let input = [ColumnRangeInput::new(interval, true, false, stats)];
    assert_eq!(
        estimate_column_row_count(&input, 100.0, 100.0, 0.0, 1.0, false, 0.01),
        RowEstimate::new(9.0, 8.0, 10.0)
    );

    let null_to_value = ScalarRange::new(0.0, 10.0, PseudoBoundKind::Null, PseudoBoundKind::Value);
    let null_stats = ColumnRangeStats::new(
        RowEstimate::default_est(0.0),
        RowEstimate::default_est(2.0),
        RowEstimate::default_est(0.0),
        RowEstimate::default_est(1.0),
        None,
    );
    let input = [ColumnRangeInput::new(
        null_to_value,
        false,
        false,
        null_stats,
    )];
    assert_eq!(
        estimate_column_row_count(&input, 100.0, 97.0, 3.0, 1.0, false, 0.01),
        RowEstimate::default_est(6.0)
    );
}

#[test]
fn partial_stats_preserve_product_and_correlated_bounds() {
    let ranges = [
        PartialStatsRange::new(vec![10.0, 20.0]),
        PartialStatsRange::new(vec![50.0, 10.0]),
    ];
    assert_eq!(
        pseudo_row_count_with_partial_stats(&ranges, 100.0, false),
        tidb_planner::cardinality::row_count_column::PartialStatsRowCount {
            total_count: 7.0,
            max_count: 20.0,
        }
    );
    assert_eq!(
        pseudo_row_count_with_partial_stats(&ranges, 100.0, true),
        tidb_planner::cardinality::row_count_column::PartialStatsRowCount {
            total_count: 60.0,
            max_count: 0.0,
        }
    );
    assert_eq!(
        pseudo_row_count_with_partial_stats(&[], 0.0, false),
        tidb_planner::cardinality::row_count_column::PartialStatsRowCount {
            total_count: 0.0,
            max_count: 0.0,
        }
    );
}
